### PR TITLE
tests: Add test case for the constraint check on repeated pessimistic lock request on left pessimistic lock

### PR DESCRIPTION
**Pull request by MyonKeminta**
Signed-off-by: MyonKeminta <MyonKeminta@users.noreply.github.com>

### What problem does this PR solve?

Issue Number: Ref #40114

Requires https://github.com/tikv/tikv/pull/14037 . The new test in this PR can only pass with TiKV after merging https://github.com/tikv/tikv/pull/14037 to TiKV.

Problem Summary:

### What is changed and how it works?

This PR adds integration test hat covers issue 40114.

### Check List

Tests 

- [x] Integration test

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by sticnarf** [comment]
This test seems to require a very recent TiKV build which contains a commit merged today. I have no idea how to use a just merged TiKV commit in CI.

File: tests/realtikvtest/pessimistictest/pessimistic_test.go [hunk 1]
```diff
@@ -3313,3 +3313,47 @@ func TestAggressiveLockingRetry(t *testing.T) {
...
+	require.NoError(t, failpoint.Enable("tikvclient/twoPCRequestBatchSizeLimit", "return"))
+	require.NoError(t, failpoint.Enable("tikvclient/beforeAsyncPessimisticRollback", `return("skip")`))
+	defer func() {
+		require.NoError(t, failpoint.Disable("tikvclient/twoPCRequestBatchSizeLimit"))
+		require.NoError(t, failpoint.Disable("tikvclient/beforeAsyncPessimisticRollback"))
+	}()
+
+	tk.MustExec("set @@innodb_lock_wait_timeout = 1")
+	tk.MustExec("begin pessimistic")
+	tk.MustExec("begin ")
```
Comment (ekexium) [inline tests/realtikvtest/pessimistictest/pessimistic_test.go:3336]:
```suggestion
	tk.MustExec("begin")
```

File: tests/realtikvtest/pessimistictest/pessimistic_test.go [hunk 2]
```diff
@@ -3313,3 +3313,47 @@ func TestAggressiveLockingRetry(t *testing.T) {
...
+	// `beforeAsyncPessimisticRollback`, which simulates a network fault).
+	// Ensure the row is still locked.
+	time.Sleep(time.Millisecond * 50)
+	tk2.MustExec("begin pessimistic")
+	tk2.MustGetErrCode("select * from t where id = 1 for update nowait", mysql.ErrLockAcquireFailAndNoWaitSet)
+	tk2.MustExec("rollback")
+
+	// tk is still in transaction.
+	tk.MustQuery("select @@tidb_current_ts = 0").Check(testkit.Rows("0"))
+	// This will unexpectedly success in issue 40114.
```
Comment (ekexium) [inline tests/realtikvtest/pessimistictest/pessimistic_test.go:3354]:
```suggestion
	// This will unexpectedly succeed in issue 40114.
```

**Review by ekexium — APPROVED** [review]

File: tests/realtikvtest/pessimistictest/pessimistic_test.go (hunk 1 above)
Comment (sticnarf) [reply tests/realtikvtest/pessimistictest/pessimistic_test.go:3336]: This line should just be removed? `tk` has started its transaction in the last line.

**Review by sticnarf — APPROVED** [review]

**Pull request merged**